### ISSUE TITLE
Add sanitizer tools to CI

### ISSUE THREAD

**Issue by tsusanka** [trezor/trezor-firmware#44]
We currently use some sanitizers on Trezor One. It'd be nice to add it to CI probably on manual invocation.

**Comment by tsusanka** [comment]
Tracked in #108.

**Comment by tsusanka** [comment]
Let's track this here after all, because it might take some time to finish this one. 

- [ ] @onvej-sl will prepare which sanitizer flags we want to use
- [ ] @tsusanka's goal is to add those to CI - if any output is rendered, the suite should fail

**Comment by onvej-sl** [comment]
tl;dr

I suggest running device tests for Trezor One and tests for trezor crypto with the following compile flags
```
-fsanitize=address -fsanitize=undefined -fno-sanitize=bool -D_FORTIFY_SOURCE=2 -fno-omit-frame-pointer -fno-optimize-sibling-calls -O1
```
and the following environment variables
```
ASAN_OPTIONS="detect_stack_use_after_return=1 strict_string_checks=1 detect_invalid_pointer_pairs=1"
```

- - - -

I made some research on sanitizers we may want to use, which are address sanitizer (ASAN), undefined behavior (USAN) sanitizer and memory sanitizer (MSAN).

[Address sanitizer](https://github.com/google/sanitizers/wiki/AddressSanitizer) detects mainly:
  * heap and stack overflow
  * memory leak
  * reference of a variable after free
  * reference of a local variable after return

It is turned on with the compile flag `-fsanitize=address` and runtime environment variable `ASAN_OPTIONS="detect_stack_use_after_return=1 strict_string_checks=1 detect_invalid_pointer_pairs=1 detect_leaks=1"`. See [this](https://github.com/google/sanitizers/wiki/AddressSanitizerFlags#run-time-flags) for details.

[Undefined behavior sanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) detect behaviour which is undefined according to the C standard, which is for example:
  * signed integer overflow
  * pointer overflow
  * reference of a null pointer

USAN is turned on with the compile flag `-fsanitize=undefined`. The flag `-fno-sanitize=bool` turns off some checks which cause a lot of false positives.

[Memory sanitizer](https://github.com/google/sanitizers/wiki/MemorySanitizer) detects use of uninicialised variables. It is turned on by the compile flags `-fsanitize=memory -fsanitize-memory-track-origins -fsanitize-memory-use-after-dtor` and runtime environment variable `MSAN_OPTIONS="poison_in_dtor=1"`.

You can use either address sanitizer or memory sanitizer, not both together. Emulator compiled with the MSAN finds a lot of leaks in `libSDL2`, trezor-crypto compiled with the MSAN finds a lot of leaks in `libcheck`. Therefore MSAN unusable for both emulator and trezor-crypto.

Note that compile flags must be passed to both compiler and linker.

Other useful flags are `-D_FORTIFY_SOURCE=2`, which adds some compile time and runtime checks of the use of string and memory manipulation functions and `-fno-omit-frame-pointer -fno-optimize-sibling-calls -O1` which might enhance usefulness of backtraces in error messages.

Here is the compile time and runtime overhead for firmware emulator and each sanitizer. Runtime overhead was measured by comparing duration of the device tests. 

| Sanitizer |  Compile  |  Runtime  |
|-----------|:---------:|:---------:|
| no        | 1         | 1         |
| ASAN      | 1.5       | 1.2       |
| USAN      | 10        | 2         |

Here is the overhead for trezor-crypto, runtime overhead was measured by the run of `test_check`.

| Sanitizer |  Compile  |  Runtime  |
|-----------|:---------:|:---------:|
| no        | 1         | 1         |
| ASAN      | 1.5       | 3         |
| USAN      | 20        | 3         |

**Comment by mmilata** [comment]
Address sanitizer via gcc added in #2052.

Still, clang might catch different errors than gcc. Also it would be nice to use clang static analysis tools `clang-tidy`/`clang-analyzer`. Currently clang+asan core emulator crashes on startup:
```
Waiting for emulator to come up...
AddressSanitizer:DEADLYSIGNAL
=================================================================
==768125==ERROR: AddressSanitizer: stack-overflow on address 0x7ffe7aa3f0b0 (pc 0x7ffe7aa3f0b0 bp 0x7ffe7aa3eef0 sp 0x7ffe7aa3eef8 T0)
    #0 0x7ffe7aa3f0b0  ([stack]+0x1f0b0)

SUMMARY: AddressSanitizer: stack-overflow ([stack]+0x1f0b0) 
==768125==ABORTING
```
